File: 74. Search a 2D Matrix.go
// https://leetcode.com/problems/search-a-2d-matrix/

package leetcode

func searchMatrix(matrix [][]int, target int) bool {
	if len(matrix) == 0 || len(matrix[0]) == 0 {
		return false
	}

	rows, cols := len(matrix), len(matrix[0])
	totalCells := rows * cols
	l, r := 0, totalCells-1

	for l <= r {
		mid := (l + r) / 2
		row, col := mid/cols, mid%cols
		val := matrix[row][col]

		switch {
		case val < target:
			l = mid + 1
		case val > target:
			r = mid - 1
		default:
			return true
		}
	}

	return false
}

File: 74. Search a 2D Matrix.js
// https://leetcode.com/problems/search-a-2d-matrix/

/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function(m, target) {
    if(!m.length || !m[0].length) return false;

    const [rows, cols] = [m.length, m[0].length];
    const totalCells = rows * cols;
    let [l, r] = [0, totalCells-1];

    while(l <= r){
        const mid = Math.floor((l+r)/2);
        const row = Math.floor(mid / cols);
        const col = mid % cols;
        const val = m[row][col];

        if(val < target) l = mid + 1;
        else if(val > target) r = mid - 1;
        else return true;
    }

    return false;
};
